modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,330 @@
+import { Optional } from './Optional';
+import * as Type from './Type';
+
+// The `Maybe` type represents a value (of any type) that potentially does not
+// exist. Any `Maybe<T>` can either be a `Just<T>` (in which case the value does
+// exist, and can be accessed as `.value`) or a `Nothing` (in which case the
+// value does not exist). This file defines a whole lot of FP-inspired utility
+// functions for dealing with Maybe objects. Things to note:
+// 1. Functions are grouped based on, primarily, their Haskell roots. That
+// said, I'm not a Haskell programmer so it might be imperfect.
+// 2. Many of the functions are "curried". This means that they take all of
+// their arguments that are not the Maybe object first, and then they return a
+// function that takes the Maybe object and actually performs the operation.
+// This is done to make the functions easier to use with Fun.pipe, so that code
+// can be written left to right.
+//
+// Comparison with null or undefined
+// - We don't get fancy null coalescing operators with Maybe
+// - We do get fancy helper functions with Maybe
+// - Maybes support nesting, and allow for the type to still be nullable (or
+// another Maybe)
+// - There is no option to turn of strict-maybe-checks like there is for
+// strict-null-checks
+// - Try to use Maybe instead of null or undefined where you can
+//
+// Comparison with Optional
+// - Optional uses a.method().method(), Maybe uses Fun.pipe(a, method(), method())
+// - Optional is invariant, Maybe is covariant (Maybe<Dog> counts as a
+// Maybe<Animal> but Optional<Dog> does not count as Optional<Animal>)
+// - Maybe tree-shakes and minifies a lot better
+// - Maybe can be `console.log`ed
+// - Maybe might perform better because it doesn't need to create messy objects
+// - Optional is deprecated (but still in very wide use) and you should try to
+// migrate over to Maybe where you can
+
+// --- Types ---
+
+// Technically, this will compile down to 0 for Nothing, and 1 for Just. If
+// you're logging a `Maybe<T>` and you see some 1s and 0s where you expected to
+// see useful strings, this is why. Leaving them as numbers comes with two
+// advantages:
+// 1. Makes me feel like a C programmer again, using 0 to mean false
```

Comment:
This comment could be confusing to others later on, as "me" is only applicable while reviewing this code or while we know who the author is.